notes/thread-design.md

```diff
@@ -0,0 +1,142 @@
+            % Points libfdb_c to the directory containing the dynamic
+            % library(ies) to be loaded. libfdb_c will copy each .so N times
+            % to temporary files on the filesystem so that they can be
+            % individually loaded. Normally the value is a binary string of
+            % the directory on the filesystem containing the dynamic
+            % library(ies).
+            %
+            % You may also provide an `{M, F, A}` tuple. If you do, the function
+            % is executed at the time the `erlfdb_nif` module is loaded. The
+            % return must be a binary string. This value is used as the
+            % permanent value for `external_client_directory` for the lifetime
+            % of the VM.
+            {external_client_directory, {erlfdb_network_options, compile_time_external_client_directory, []}},
+
+            % `client_threads_per_version`:
+            % Number of threads to create per dynamic library. This is the
+            % value to tweak in order to scale the number of client threads
+            % horizontally. This value is used as the permanent value
+            % for the lifetime of the BEAM VM.
+            %
+            % When the value is 1, only the local client is used. When the
+            % value is N > 1, then N external client threads are used.
+            %
+            % You may also provide an `{M, F, A}` tuple. If you do, the
+            % function is executed at the time the `erlfdb_nif` module is
+            % loaded. The return must be a positive integer.
+            %
+            % Note: For each thread, libfdb_c will create a tcp
+            % connection for each coordinator in the cluster file.
+            {client_threads_per_version, 1}
+        ]}
+    ]}
+].
+```
+
+When the erlfdb NIF is loaded, any network_options defined by your config are merged into the defaults, so that yours always take precedence. The final set of resolved network_options is stored in env var `network_options_resolved`.
+
+```erlang
+application:get_env(erlfdb, network_options_resolved).
+```
+
+### How to choose the correct value for `client_threads_per_version`
+
+The ideal approach is to choose an integer value that is sufficient for your workload, and no more. It's not recommended to choose a value that is larger than the number of online schedulers.
+
+Instead of choosing a specific value, you may wish for erlfdb to scale dynamically along with your scaling of the Erlang VM itself. Consider the following MFA-tuple choices.
+
+- `{erlang, system_info, [schedulers_online]}`
+- `{erlang, system_info, [dirty_io_schedulers]}`
+
+> ** Note **: These suggestions are made simply as a convenience for automatic scaling, and none of these choices will associate erlfdb external client threads with Erlang's schdeulers or dirty_io_schedulers. External client threads will always be created **in addition to** and **not replacements for** any of the Erlang VM's OS threads.
```

Comment:
Does Erlang give us any control over which CPU the external client thread runs on? If so, could we use that to ensure the scheduler always dispatches work to the external client thread on the same CPU? And would that help us in any way? (Or potentially harm us, if the scheduler and external client thread were fighting for resources?)

Reply:
No, Erlang is not actually aware of these threads at all, since they are started by libfdb_c itself. I don't believe there's any way to control the affinity, at least without some OS specific operations.